### PR TITLE
New: Updated Rarbg request limits

### LINKED ISSUE #5206: TorrentAPI new limits/changes
### Is there an existing issue for this?

- [X] I have searched the existing issues

### Is your feature request related to a problem? Please describe

--

### Describe the solution you'd like

No clue who's responsible for Sonarr right now but @bakerboy448 said to open an issue so here is it

To fix Sonarr potential issues please implement the following in your torrentapi code :

1. Increase the sleep between requests x2
2. if the API returns 429 retry in 2 minutes , 429 returns a custom json with rate_limit 1 - ignore it
3. if the API returns 520 retry in 3 minutes
4. if the API returns 200 with a json that contains rate_limit 1 retry in 5 minutes

That will cover all cases. Back-off is unnecessary if it really waits for 6 hours. I never experienced a back-off period in my Sonarr so far tho.

### Describe alternatives you've considered

--

### Anything else?

--

### PR DESCRIPTION
**Pull request by markus101**
#### Database Migration
NO

#### Description
Updates Rarbg to handle additional rate limit cases.

#### Todos
- [x] Tests

#### Issues Fixed or Closed by this PR

* Closes #5206

### CODE REVIEW

File: src/NzbDrone.Core/Indexers/Rarbg/RarbgParser.cs [hunk 1]
```diff
@@ -28,6 +33,11 @@ public IList<ReleaseInfo> ParseResponse(IndexerResponse indexerResponse)
 
             var jsonResponse = new HttpResponse<RarbgResponse>(indexerResponse.HttpResponse);
 
+            if (jsonResponse.Resource.rate_limit == 1)
```
Comment (bakerboy448) [inline src/NzbDrone.Core/Indexers/Rarbg/RarbgParser.cs:36]: remove this for now.  RarBG has a bug where rate_limit = 1 on almost every request

Ref - https://github.com/Prowlarr/Prowlarr/issues/1277
Ref - https://github.com/Prowlarr/Prowlarr/commit/5cc044aa8fbc7ba62d9c32442d8125b9b347d28a

**Comment by bakerboy448** [comment]
would it be prudent to remove the defunct cloudflare /captcha logic and associated settings as well?

File: src/NzbDrone.Core/Indexers/Rarbg/RarbgParser.cs [hunk 2]
```diff
@@ -28,6 +33,12 @@ public IList<ReleaseInfo> ParseResponse(IndexerResponse indexerResponse)
 
             var jsonResponse = new HttpResponse<RarbgResponse>(indexerResponse.HttpResponse);
 
+            // TODO: Uncomment when RARBG doesn't return it for most valid requests
+            // if (jsonResponse.Resource.rate_limit == 1)
```
Comment (bakerboy448) [inline src/NzbDrone.Core/Indexers/Rarbg/RarbgParser.cs:37]:
```suggestion
            if (jsonResponse.Resource.rate_limit is 1 && jsonResponse.Resource.torrent_results == null)
```

Comment (bakerboy448) [reply src/NzbDrone.Core/Indexers/Rarbg/RarbgParser.cs:37]: uncomment & only throw if there are no results to workaround false positive

File: src/NzbDrone.Core/Indexers/Rarbg/RarbgParser.cs [hunk 3]
```diff
@@ -44,6 +49,11 @@ public IList<ReleaseInfo> ParseResponse(IndexerResponse indexerResponse)
 
             if (jsonResponse.Resource.torrent_results == null)
             {
+                if (jsonResponse.Resource.rate_limit == 1)
```
Comment (bakerboy448) [inline src/NzbDrone.Core/Indexers/Rarbg/RarbgParser.cs:52]: suggest skip handling jsonResponse.Resource.error_code == 5 above and handle here instead
5 indicates a rate limit

Comment (markus101) [reply src/NzbDrone.Core/Indexers/Rarbg/RarbgParser.cs:52]: I’m basing the changes on the GHI, if we need to make other changes later we can.

**Comment by bakerboy448** [comment]
Based on comments / users with jackett's testing - it probably makes sense to comment out the the rate_limit = 1 handling and land this otherwise as is.

x-ref https://github.com/Jackett/Jackett/pull/14000#issuecomment-1426925111

**Comment by bakerboy448** [comment]
FYI some prowlarr rework to catch the http error before it tries to parse so the html isn't tried to be parsed

https://github.com/Prowlarr/Prowlarr/pull/1458

File: src/NzbDrone.Core/Indexers/Rarbg/RarbgParser.cs [hunk 4]
```diff
@@ -17,6 +18,10 @@ public IList<ReleaseInfo> ParseResponse(IndexerResponse indexerResponse)
 
             switch (indexerResponse.HttpResponse.StatusCode)
             {
+                case HttpStatusCode.TooManyRequests:
+                    throw new RequestLimitReachedException("Indexer API limit reached", TimeSpan.FromMinutes(2));
+                case (HttpStatusCode)520:
+                    throw new RequestLimitReachedException("Indexer API returned unknown error", TimeSpan.FromMinutes(3));
```
Comment (bakerboy448) [inline src/NzbDrone.Core/Indexers/Rarbg/RarbgParser.cs:24]: Nit: 520 is known - internal servers behind cloudflare are rate limited/overwhelmed

```
when rated limited web returns http code 444 to cloudflare which returns 520 on cloudflare , there is no way to send 429 to cloudflare without it breaking the request properly and retrying it
```

Comment (markus101) [reply src/NzbDrone.Core/Indexers/Rarbg/RarbgParser.cs:24]: Unless it throws a 520 for another reason…

**Comment by yardenshoham** [comment]
When will this be released?

**Comment by bakerboy448** [comment]
Please use one of the support channels: [forums](https://forums.sonarr.tv/), [subreddit](https://www.reddit.com/r/sonarr/), [discord ](https://discord.gg/Ex7FmFK), or [IRC ](https://web.libera.chat/?channels=#sonarr)for support/questions.

no eta on when v4 beta would be released

also a moot point as rarbg is defunct

**Pull request merged**